v3_async_wip/v3_async_wip/iothub_session.py

```diff
@@ -0,0 +1,284 @@
+                "Incompatible authentication - cannot provide both 'shared_access_key' and 'sastoken_fn'"
+            )
+        if not shared_access_key and not sastoken_fn and not ssl_context:
+            raise ValueError(
+                "Missing authentication - must provide one of 'shared_access_key', 'sastoken_fn' or 'ssl_context'"
+            )
+
+        # Set up SAS auth (if using)
+        generator: Optional[st.SasTokenGenerator]
+        # NOTE: Need to keep a reference to the SasTokenProvider so we can stop it during cleanup
+        self._sastoken_provider: Optional[st.SasTokenProvider]
+        if shared_access_key:
+            uri = _format_sas_uri(hostname=hostname, device_id=device_id, module_id=module_id)
+            signing_mechanism = sm.SymmetricKeySigningMechanism(shared_access_key)
+            generator = st.InternalSasTokenGenerator(signing_mechanism=signing_mechanism, uri=uri)
+            self._sastoken_provider = st.SasTokenProvider(generator)
+        elif sastoken_fn:
+            generator = st.ExternalSasTokenGenerator(sastoken_fn)
+            self._sastoken_provider = st.SasTokenProvider(generator)
+        else:
+            self._sastoken_provider = None
+
+        # Create a default SSLContext if not provided
+        if not ssl_context:
+            ssl_context = _default_ssl_context()
+
+        # Instantiate the MQTTClient
+        cfg = config.IoTHubClientConfig(
+            hostname=hostname,
+            device_id=device_id,
+            module_id=module_id,
+            sastoken_provider=self._sastoken_provider,
+            ssl_context=ssl_context,
+            auto_reconnect=False,  # No reconnect for now
+            **kwargs,
+        )
+        self._mqtt_client = mqtt.IoTHubMQTTClient(cfg)
+
+    async def __aenter__(self) -> "IoTHubSession":
+        # First, if using SAS auth, start up the provider
+        if self._sastoken_provider:
+            # NOTE: No try/except block is needed here because in the case of failure there is not
+            # yet anything that we would need to clean up.
+            await self._sastoken_provider.start()
+
+        # Start/connect
+        try:
+            await self._mqtt_client.start()
+            await self._mqtt_client.connect()
+        except (Exception, asyncio.CancelledError):
+            # Stop/cleanup if something goes wrong
+            await self._stop_all()
+            raise
+
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        traceback: TracebackType,
+    ) -> None:
+        # NOTE: these are not supposed to be able to fail so no error handling required
+        await self._mqtt_client.disconnect()
+        await self._stop_all()
+
+    async def _stop_all(self) -> None:
+        await self._mqtt_client.stop()
+        if self._sastoken_provider:
+            await self._sastoken_provider.stop()
```

Comment:
How about this?
```
try:
  await self._mqtt_client.stop()
finally:
  if self._sastoken_provider:
    await self._sastoken_provider.stop()
```
This way the sastoken provider stops even if there's an exception stopping the client. (Using `try`/`finally` instead of re-raising ensures that the original exception context doesn't get lost.)

Reply:
Added this and detailed tests